### PR TITLE
draft validation functions

### LINKED ISSUE #94: Ensure CI is set up correctly and runs tests

### LINKED ISSUE #97: validation function checking that all files exist
This validation function will check for the presence of all expected top-level files.

### LINKED ISSUE #100: validation function checking that root mesh and scaled annotations image roughly match
This consists of calculating the axis-aligned bounding box (AABB) of the labels in `annotation.tiff` and of the root mesh, scaling the AABB of the labels and checking that the two are closer than a threshold - the threshold could depend on the pixel size. Just needs to enable us to catch things like #86 .

### PR DESCRIPTION
**Pull request by viktorpm**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

We want to test if all the atlases fit the brainglobe framework

**What does this PR do?**

* Adds validation function for the existence of important files
* Adds validation function for mesh and annotation size (roughly the same)
* Adds a script that loops over all atlases, updates and validates them 
* adds initial tests for these validation functions (one positive and one negative unit test for each)
* drops python 3.8 support for GH actions and tox, and ensures pytest is run in tox.
* suggests an initial structure for validation functionality

## References

gets started with #81 
also fixes #94 
fixes #100 
fixes #97 

## How has this PR been tested?

It was run on the HPC. We got the expected output: two lists with passed and failed atlases
 
```
Summary
['example_mouse_100um', 'allen_mouse_10um', 'allen_mouse_25um', 'allen_mouse_50um', 'allen_mouse_100um', 'kim_mouse_50um', 'osten_mouse_10um', 'osten_mouse_25um', 'osten_mouse_50um', 'osten_mouse_100um']
[('mpin_zfish_1um', AssertionError('Mesh coordinate 962.676331 and annotation coordinate 872.0 differ by more than 10 times pixel size 1.0')), ('allen_human_500um', AssertionError('Mesh coordinate 232583.5 and annotation coordinate 24500.0 differ by more than 10 times pixel size 500.0')), ('kim_mouse_10um', AssertionError('Mesh coordinate 175.0 and annotation coordinate 860.0 differ by more than 10 times pixel size 10.0')), ('kim_mouse_25um', AssertionError('Mesh coordinate 437.5 and annotation coordinate 875.0 differ by more than 10 times pixel size 25.0')), ('kim_mouse_100um', AssertionError('Mesh coordinate 25550.0 and annotation coordinate 13000.0 differ by more than 10 times pixel size 100.0')), ('allen_cord_20um', AssertionError('Mesh coordinate 390.0 and annotation coordinate 0.0 differ by more than 10 times pixel size 20.0')), ('azba_zfish_4um', AssertionError('Mesh coordinate 59.675 and annotation coordinate 244.0 differ by more than 10 times pixel size 4.0')), ('whs_sd_rat_39um', AssertionError('Mesh coordinate 955.5 and annotation coordinate 0.0 differ by more than 10 times pixel size 39.0')), ('perens_lsfm_mouse_20um', AssertionError('Mesh coordinate 120209.2 and annotation coordinate 120.0 differ by more than 10 times pixel size 20.0')), ('admba_3d_e11_5_mouse_16um', AssertionError('Mesh coordinate 2136.0 and annotation coordinate 5088.0 differ by more than 10 times pixel size 16.0')), ('admba_3d_e13_5_mouse_16um', AssertionError('Mesh coordinate 2648.0 and annotation coordinate 6160.0 differ by more than 10 times pixel size 16.0')), ('admba_3d_e15_5_mouse_16um', AssertionError('Mesh coordinate 3768.0 and annotation coordinate 7296.0 differ by more than 10 times pixel size 16.0')), ('admba_3d_e18_5_mouse_16um', AssertionError('Mesh coordinate 120.0 and annotation coordinate 400.0 differ by more than 10 times pixel size 16.0')), ('admba_3d_p4_mouse_16.752um', AssertionError('Mesh coordinate 6307.128 and annotation coordinate 11240.591999999999 differ by more than 10 times pixel size 16.752')), ('admba_3d_p14_mouse_16.752um', AssertionError('Mesh coordinate 6374.1359999999995 and annotation coordinate 13016.303999999998 differ by more than 10 times pixel size 16.752')), ('admba_3d_p28_mouse_16.752um', AssertionError('Mesh coordinate 6558.407999999999 and annotation coordinate 14205.696 differ by more than 10 times pixel size 16.752')), ('admba_3d_p56_mouse_25um', AssertionError('Mesh coordinate 437.5 and annotation coordinate 0.0 differ by more than 10 times pixel size 25.0')), ('princeton_mouse_20um', AssertionError('Mesh coordinate 618.5 and annotation coordinate 12360.0 differ by more than 10 times pixel size 20.0')), ('kim_dev_mouse_stp_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_idisco_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_mri_a0_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_mri_adc_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_mri_dwi_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_mri_fa_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_mri_mtr_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0')), ('kim_dev_mouse_mri_t2_10um', AssertionError('Mesh coordinate 515.0 and annotation coordinate 30.0 differ by more than 10 times pixel size 10.0'))]
```

## Is this a breaking change?

No

## Does this PR require an update to the documentation?

Not for now. 

## Checklist:

- [NA] The code has been tested locally 
- [NA] Tests have been added to cover all new functionality (unit & integration)
- [NA] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: bg_atlasgen/validate_atlases.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    updated = get_atlases_lastversions()[atlas_name]["updated"]
+    if not updated:
+        update_atlas(atlas_name)
+    atlas_path = Path(get_brainglobe_dir()) / f"{atlas_name}_v{version}"
+    assert validate_atlas_files(
+        atlas_path
+    ), f"Atlas file {atlas_path} validation failed"
+    assert validate_mesh_matches_image_extents(
+        atlas
+    ), "Atlas object validation failed"
```
Comment (niksirbi) [inline bg_atlasgen/validate_atlases.py:106]: "Atlas object validation" is too abstract for this message. You didn't check the entire object, just the extents of the annotation image and the mesh, right?

Comment (niksirbi) [reply bg_atlasgen/validate_atlases.py:106]: Unless this message anticipates the addition of more checks in this group

**Review by niksirbi — APPROVED** [review]
Great job @viktorpm and @alessandrofelder! 

The validation functions appear to do what they claim and are very undestandable. 

I've sprinkled several pedantic comments throughout, but feel free to take them or leave them, as many of them are a matter of style.

Now coming to my only substantive comment, which concers the capacity of this design to accommodate multiple future validation functions.

In the `validate_atlas()` funciton, the current two validations are performed sequentially, i.e. first `assert validate_atlas_files()` and then `assert validate_mesh_matches_image_extents()`.

In this case, it makes sense to have them run sequentially, since there is no point in looking at the image extents if the files are not there to begin with.

However, in future, we may add validators that are not dependent on each other, for example: 
- A: validate_mesh_normals
- B: validate_annotation_image_orientation

In this case, if you assert A, then B, the first assertion will fail if there is a problem with the mesh normals, and the second assertion will never run. So you will not learn if there is also a problem with image orientaiton. At least that's my own prediction of what will happen in that case, correct me if I have misread the code.

It would be nice to have a report of all the independent problems with an atlas, not just the first problem encountered.

I wonder if you guys have thought about such use cases - i.e. how to run multiple independent assertions. One solution could be catching each "AssertionError" and logging it into a list of errors instead of erroring out outright. But I'm sure there are also other ways. Since this problem concerns future validators, and not the curently existing ones, feel free to open an issue instead of fixing here (if you prefer).

**P.S:** This PR contains updates to Python tooling, like the dev dependencies and the gh actions. Normally, I'd say it's best practice to implement those in an independent PR, but I understand the practical considerations in this case, so I won't complain about it.

File: bg_atlasgen/validate_atlases.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        "reference.tiff",
+        "metadata.json",
+        "structures.json",
+        "meshes",
+    ]
+    for expected_file_name in expected_files:
+        expected_path = Path(atlas_path / expected_file_name)
+        assert (
+            expected_path.is_file()
+        ), f"Expected file not found at {expected_path}"
```
[the remark this answers is not in the document]
Comment (viktorpm) [reply bg_atlasgen/validate_atlases.py:30]: Thank you! It's fixed now

**Review by niksirbi — APPROVED** [review]
I like the solution you guys came up with for running multiple validations and collecting all the results! Nice work!

I left only 4 tiny comments, 3 of which have to do with appeasing Solar Lint, so that the CI checks won't fail.

Approved :tada:

**Comment by viktorpm** [comment]
Thank you @niksirbi for reviewing it!

File: bg_atlasgen/validate_atlases.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+def validate_atlas(atlas_name, version, all_validation_functions):
+    """Validates the latest version of a given atlas"""
+
+    print(atlas_name, version)
+    BrainGlobeAtlas(atlas_name)
+    updated = get_atlases_lastversions()[atlas_name]["updated"]
+    if not updated:
+        update_atlas(atlas_name)
+    Path(get_brainglobe_dir()) / f"{atlas_name}_v{version}"
```
[the remark this answers is not in the document]
Comment (viktorpm) [reply bg_atlasgen/validate_atlases.py:113]: Done

File: bg_atlasgen/validate_atlases.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        # validate_mesh_matches_image_extents(atlas: BrainGlobeAtlas)
+        (BrainGlobeAtlas(atlas_name),),
+        # open_for_visual_check()
+        (),
+        # validate_checksum()
+        (),
+        # check_additional_references()
+        (),
+        # validate_atlas(atlas_name, version)
+        (atlas_name, version),
```
[the remark this answers is not in the document]
Comment (viktorpm) [reply bg_atlasgen/validate_atlases.py:127]: It was left there by mistake. Removed now

**Review by alessandrofelder — APPROVED** [review]

**Pull request merged**